[Inductor] Allow statically analyzing epilogues

PaulZhang12 opened the pull request:
Differential Revision: D88786360

Now in the case of epilogue_fusion, we can have MultiTemplateBuffers that aren't necessarily returned from benchmark_epilogue_fusion due to the advent of Async-Pipelined Autotuning. In that case, we don't always want to default to benchmarking epilogues due to compile time overhead. We enforce some level of  safety in checking that the fusion compiles/has minimal register spillage and then take the fusion.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

eellison reviewed (approved):
Looks good, my one concern is on concurrent gpu use while we're benchmarking elsewhere.

PaulZhang12 committed:
[Inductor] Allow statically analyzing epilogues (#170001)

Summary:

Now in the case of epilogue_fusion, we can have MultiTemplateBuffers that aren't necessarily returned from `benchmark_epilogue_fusion` due to the advent of Async-Pipelined Autotuning. In that case, we don't always want to default to benchmarking epilogues due to compile time overhead. We enforce some level of  safety in checking that the fusion compiles/has minimal register spillage and then take the fusion.

Test Plan: buck2 run mode/opt caffe2/test/inductor:max_autotune -- -r test_template_bad_epilogue_fusion

Reviewed By: eellison

Differential Revision: D88786360

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -2706,6 +2706,19 @@ def _replace_operation_buffer(
+def _estimate_fused_epilogue_runtime(node1, node2, epilogue_runtime) -> float:
+    # If no extra memory read by epilogue, assume epilogue is free
+    # if extra memory is read by epilogue, add to minimum choice
+    total_read_bytes = node2.get_read_buffer_sizes()
```
Will include node2 writes as a followup + overall cap on memory for not performing a fusion

eellison reviewed (approved):
can add more logic as follow ups, not enabled by default yet

malfet commented:
@pytorchbot merge -f "Was landed internally"
